Mathlib/Combinatorics/SimpleGraph/LapMatrix.lean

```diff
@@ -0,0 +1,331 @@
+  rw [←half_add_self (∑ x_1 : V, ∑ x_2 : V, _)]
+  conv_lhs => arg 1; arg 2; arg 2; intro i; arg 2; intro j; rw [if_congr (adj_comm G i j) rfl rfl]
+  conv_lhs => arg 1; arg 2; rw [Finset.sum_comm]
+  simp [← sum_add_distrib]
+  conv_lhs => arg 1; arg 2; intro i; arg 2; intro j; rw [ite_add_ite]
+  field_simp
+  rw [sum_congr]
+  · rfl
+  · intros i _
+    rw [sum_congr]
+    · rfl
+    · intros j _
+      split
+      · rw [pow_two]
+        ring
+      · rfl
+
+/-- The Laplacian matrix is positive semidefinite -/
+theorem isPosSemidef_lapMatrix : (G.lapMatrix ℝ).PosSemidef := by
+  unfold PosSemidef
+  constructor
+  · unfold IsHermitian; rw [conjTranspose_eq_transpose_of_trivial, isSymm_lapMatrix]
+  · intro x
+    rw [star_trivial, ← toLinearMap₂'_apply', lapMatrix_toLinearMap₂']
+    apply div_nonneg_iff.mpr; left; constructor
+    · apply sum_nonneg'
+      intro i
+      apply sum_nonneg'
+      intro j
+      split_ifs
+      · exact sq_nonneg (x i - x j)
+      · rfl
+    · exact zero_le_two
+
+/-- `sqrtDiagMatrix A` is the diagonal matrix consisting of the square root of the diagonal
+  entries of `A`. The square roots of negative entries are set to zero. -/
+noncomputable def sqrtDiagMatrix (A : Matrix V V ℝ) : Matrix V V ℝ :=
+    Matrix.diagonal (λ i ↦ Real.sqrt (Matrix.diag A i))
+
+/-- If `A` is a diagonal matrix with nonnegative entries, `sqrtDiagMatrix A` multiplied
+  by its transpose is `A`. -/
+theorem sqrt_diag_matrix_square (A : Matrix V V ℝ) (h : IsDiag A) (h' : ∀ i : V, 0 ≤ A i i) :
+    (sqrtDiagMatrix A)ᵀ * sqrtDiagMatrix A = A := by
+  ext i j
+  simp only [sqrtDiagMatrix, diag_apply, diagonal_transpose, mul_apply, ne_eq, diagonal_apply,
+    mul_ite, ite_mul, zero_mul, mul_zero, sum_ite_eq', mem_univ, ite_true]
+  split_ifs with hij
+  · rw [hij, Real.mul_self_sqrt]
+    exact h' j
+  · rw [← h]
+    exact hij
+
+/-- The square root of the diagonal of a positive semidefinite matrix-/
+noncomputable def sqrtDiagMatrix_eigenvectorMatrix {A : Matrix V V ℝ} (hA : PosSemidef A) :=
+   (sqrtDiagMatrix (diagonal ((↑) ∘ hA.1.eigenvalues)) * hA.1.eigenvectorMatrixᵀ)
+
+lemma PosSemidef.spectral_theorem {A : Matrix V V ℝ} (hA : PosSemidef A) :
+    A = (sqrtDiagMatrix_eigenvectorMatrix hA)ᵀ * (sqrtDiagMatrix_eigenvectorMatrix hA) := by
+  unfold sqrtDiagMatrix_eigenvectorMatrix
+  rw [transpose_mul, transpose_transpose, mul_assoc]
+  conv => rhs; arg 2; rw [← mul_assoc];
+  rw [sqrt_diag_matrix_square, ← conjTranspose_eq_transpose_of_trivial,
+    Matrix.IsHermitian.conjTranspose_eigenvectorMatrix hA.1]
+  · rw [← mul_assoc]
+    simp [Matrix.IsHermitian.spectral_theorem' hA.1]
+    rfl
+  · simp only [isDiag_diagonal]
+  · intro i
+    rw [diagonal_apply_eq]
+    apply PosSemidef.eigenvalues_nonneg
+    exact hA
+
+theorem PosSemidef.toLinearMap₂'_zero_iff (A : Matrix V V ℝ) (h_psd : PosSemidef A) (x : V → ℝ) :
+    Matrix.toLinearMap₂' A x x = 0 ↔ Matrix.toLinearMap₂' A x = 0 := by
```

Comment:
I think it would be cleaner to express this in terms of `Matrix.toLin'` (i.e. interpreting A as a linear, not bilinear, map). Can you show
```
theorem PosSemidef.toLin'_zero_iff (A : Matrix V V ℝ) (h_psd : PosSemidef A) (x : V → ℝ) :
    Matrix.toLinearMap₂' A x x = 0 ↔ Matrix.toLin' A x = 0 := by sorry
```
? Then you can replace most of the uses of `toLinearMap₂'` (the ones which aren't explicitly using the bilinear form) further down the file with `toLin'`.